src/modules/iotjs_module_mqtt.c

```diff
@@ -565,8 +587,175 @@ JS_FUNCTION(MqttHandle) {
+    case MQTT_ERR_BAD_IDENTIFIER:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: bad client identifier");
+    case MQTT_ERR_SERVER_UNAVIABLE:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: server unavailable");
+    case MQTT_ERR_BAD_CREDENTIALS:
       return JS_CREATE_ERROR(
-          COMMON, "MQTT: Unallowed packet has arrived to the client");
+          COMMON, "MQTT: Connection refused: bad username or password");
+    case MQTT_ERR_UNAUTHORISED:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Connection refused: unauthorised");
+    case MQTT_ERR_CORRUPTED_PACKET:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Corrupted packet");
+    case MQTT_ERR_UNALLOWED_PACKET:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Broker sent an unallowed packet");
+    case MQTT_ERR_SUBSCRIPTION_FAILED:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Subscription failed");
+    default:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Unknown error");
+  }
+}
+
+
+JS_FUNCTION(MqttReceive) {
+  DJS_CHECK_THIS();
+  DJS_CHECK_ARGS(2, object, object);
+
+  jerry_value_t jnat = JS_GET_ARG(0, object);
+
+  void *native_p;
+  JNativeInfoType *out_native_info;
+  bool has_p =
+      jerry_get_object_native_pointer(jnat, &native_p, &out_native_info);
+  if (!has_p || out_native_info != &mqttclient_native_info) {
+    return JS_CREATE_ERROR(COMMON, "MQTT native pointer not available");
+  }
+  iotjs_mqttclient_t *mqttclient = (iotjs_mqttclient_t *)native_p;
+
+  jerry_value_t jbuffer = JS_GET_ARG(1, object);
+  iotjs_bufferwrap_t *buff_recv = iotjs_bufferwrap_from_jbuffer(jbuffer);
+  IOTJS_ASSERT(buff_recv->length != 0);
+
+  // Concat the buffers if we previously needed to
+  if (mqttclient->must_concat) {
+    iotjs_mqtt_concat_buffers(mqttclient, buff_recv);
+    mqttclient->must_concat = false;
+  }
+
+  // Only 1 byte arrived, the local buffer is empty, we can't get remaining
+  // length, just concat the receive buffer on the next packet
+  if (!mqttclient->buffer_length && buff_recv->length == 1) {
+    mqttclient->must_concat = true;
+    return jerry_create_undefined();
+  }
+
+  uint32_t packet_size = 0;
+  packet_size = iotjs_decode_remaining_length(buff_recv->buffer + 1, NULL) + 2;
+  if (!packet_size) {
+    // If the remaining length has still not arrived fully, we must concat again
+    // with the next receive packet
+    mqttclient->must_concat = true;
+    return jerry_create_undefined();
+  }
+
+  // Keep on going, if data remains just continue looping
+  while (mqttclient->buffer_length + buff_recv->length >= packet_size) {
+    int ret_val = 0;
+    jerry_value_t jdata = iotjs_bufferwrap_create_buffer(packet_size);
+    iotjs_bufferwrap_t *data = iotjs_bufferwrap_from_jbuffer(jdata);
+
+    if (!mqttclient->buffer_length) {
+      memcpy(data->buffer, buff_recv->buffer, packet_size);
+      if ((ret_val = iotjs_mqtt_handle(jnat, data))) {
```

Comment:
Cannot we directly pass an uint8_t* pointer and a size rather than creating a buffer?